collectors/jiraffe/tests/test_collectors.py

```diff
@@ -396,3 +406,66 @@ def test_collect(self, pin_envs, project_key, fields_count):
 
         project_fields = JiraProjectFields.objects.filter(project_key=project_key)
         assert len(project_fields) == fields_count
+
+    @freeze_time(timezone.datetime(2015, 12, 12))
+    @pytest.mark.vcr(*BASE_METADATA_COLLECTOR_VCRS)
+    @pytest.mark.parametrize("project_key,fields_count", [("RHEL", 120), ("OSIM", 20)])
+    def test_metadata_not_deleted_after_failure(
+        self, monkeypatch, project_key, fields_count
+    ):
+        ps_module = PsModuleFactory(
+            bts_name="jira",
+            bts_key=project_key,
+            supported_until_dt=timezone.make_aware(timezone.datetime(2020, 12, 12)),
+        )
+        PsUpdateStreamFactory(ps_module=ps_module)
+
+        project_fields = JiraProjectFields.objects.filter(project_key=project_key)
+        assert len(project_fields) == 0
+
+        mc = MetadataCollector()
+        mc.collect()
+
+        project_fields = JiraProjectFields.objects.filter(project_key=project_key)
+        assert len(project_fields) == fields_count
+
+        def raiseJiraError(*args, **kwargs):
+            raise JIRAError
+
+        monkeypatch.setattr(mc.jira_querier.jira_conn, "_get_json", raiseJiraError)
+        with pytest.raises(MetadataCollectorInsufficientDataJiraffeException):
+            mc.collect()
+
+        project_fields = JiraProjectFields.objects.filter(project_key=project_key)
+        assert len(project_fields) == fields_count
+
+    @freeze_time(timezone.datetime(2015, 12, 12))
+    @pytest.mark.vcr(*BASE_METADATA_COLLECTOR_VCRS)
+    @pytest.mark.parametrize("project_key,fields_count", [("RHEL", 120), ("OSIM", 20)])
+    def test_metadata_deleted_based_on_product_definitions(
+        self, project_key, fields_count
+    ):
+        ps_module = PsModuleFactory(
+            bts_name="jira",
+            bts_key=project_key,
+            supported_until_dt=timezone.make_aware(timezone.datetime(2020, 12, 12)),
+        )
+        PsUpdateStreamFactory(ps_module=ps_module)
+
+        project_fields = JiraProjectFields.objects.filter(project_key=project_key)
+        assert len(project_fields) == 0
+
+        mc = MetadataCollector()
+        mc.collect()
+
+        project_fields = JiraProjectFields.objects.filter(project_key=project_key)
+        assert len(project_fields) == fields_count
+
+        # Remove active PS Update Streams to rule out PS Module and force project
+        # removal in Jira metadata
+        ps_module.active_ps_update_streams.set([])
```

Comment:
I am unsure about this line. I would expect this will be done automatically by code, but my understanding might be incorrect.

Reply:
Sorry, I don't understand. How the code would automatically remove active PS update streams?

Reply:
Discussed personally, this simulates a change in product definitions.